### PR TITLE
New mock test framework for windows kubeproxy.

### LINKED ISSUE #120200: Remove the dependency of HNS in Winkernel KubeProxy Tests by replacing with mock calls.
### What happened?

The current tests in WinKernel Kube proxy are dependent on Windows HNS environment. The tests expect HNS to be present and are executed by making real system calls to test environment. This makes testing difficult for the developer to write make test cases as it mandates to have a real Windows Node environment.

### What did you expect to happen?

The proposal is to have a framework which can mock all HCN calls.

### How can we reproduce it (as minimally and precisely as possible)?

Running the testcases in in proxier_test.go in development environment can help to repro the issue.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# 1.29, 1.28
```

</details>

### Cloud provider

<details>
AKS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by princepereira**
#### What type of PR is this?
/kind cleanup
/kind failing-test
/kind windows
/kind kube-proxy

#### What this PR does / why we need it:
This is to add mock framework for unit tests for windows in kube-proxy

#### Which issue(s) this PR fixes:

Fixes #120200

#### Does this PR introduce a user-facing change?
NONE

```release-note
This contains the modified windows kubeproxy testcases with mock implementation
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/proxy/winkernel/proxier.go [hunk 1]
```diff
@@ -152,11 +152,13 @@ const (
 	MAX_COUNT_STALE_LOADBALANCERS = 20
 )
```
Comment (daschott) [inline pkg/proxy/winkernel/proxier.go:153]: There are 2 methods in proxier.go that I believe are still calling hcsshim 
  * deleteAllHnsLoadBalancerPolicy() 
  * getHnsNetworkInfo() 
   
  Can we update these to also use HcnService interface methods?

**Review by daschott — COMMENTED** [review]
Great to see this, thank you for working on this

File: pkg/proxy/winkernel/hns.go [hunk 2]
```diff
@@ -31,6 +31,7 @@ import (
 )
```
[the remark this answers is not in the document]
Comment (princepereira) [reply pkg/proxy/winkernel/hns.go]: hcn.go modified to hcnutils.go

File: pkg/proxy/winkernel/hcn_mock.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
+//go:build windows
+// +build windows
+
+/*
```
[the remark this answers is not in the document]
Comment (princepereira) [reply pkg/proxy/winkernel/hcn_mock.go]: Moved hcnutils_mock to testing.

File: pkg/proxy/winkernel/hcn_mock.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+		}
+		policies = append(policies, policy)
+	}
+
+	network := &hcn.HostComputeNetwork{
+		Id:       networkInfo.id,
+		Name:     networkInfo.name,
+		Type:     hcn.NetworkType(networkInfo.networkType),
+		Policies: policies,
+	}
```
[the remark this answers is not in the document]
Comment (princepereira) [reply pkg/proxy/winkernel/hcn_mock.go]: Moved to a new function

File: pkg/proxy/winkernel/hns.go [hunk 5]
```diff
@@ -31,6 +31,7 @@ import (
...
 
 type HostNetworkService interface {
+	getHcn() HcnService
```
[the remark this answers is not in the document]
Comment (princepereira) [reply pkg/proxy/winkernel/hns.go]: Sure. Simplified the calls as per the comment. Also made hcn as part of proxier to make "getSupportedFeatures" call directly from proxier methods.

File: pkg/proxy/winkernel/proxier.go (hunk 1 above)
Comment (princepereira) [reply pkg/proxy/winkernel/proxier.go:153]: Addressed. Func: getHnsNetworkInfo() is removed as it is never used

File: pkg/proxy/winkernel/proxier_test.go [hunk 6]
```diff
@@ -20,39 +20,72 @@ limitations under the License.
...
 
+	"github.com/Microsoft/hcsshim/hcn"
 	v1 "k8s.io/api/core/v1"
 	discovery "k8s.io/api/discovery/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/kubernetes/pkg/proxy"
 	"k8s.io/kubernetes/pkg/proxy/healthcheck"
+	fakeHcn "k8s.io/kubernetes/pkg/proxy/winkernel/testing"
```
Comment (pacoxu) [inline pkg/proxy/winkernel/proxier_test.go:38]:
```
pkg/proxy/winkernel/proxier_test.go:	fakeHcn "k8s.io/kubernetes/pkg/proxy/winkernel/testing"
!!! Some package aliases break go conventions.
To fix these errors, do not use capitalized or underlined characters
```
/hold

The CI failure should be fixed. Feel free to unhold once this is fixed.

Comment (princepereira) [reply pkg/proxy/winkernel/proxier_test.go]: Issue with package name convention fixed.

/unhold

**Pull request merged**